### PR TITLE
fix(volume): Generate volume name compliant with DNS Label names

### LINKED ISSUE #961: --mount flag in update command failed with more than one level of directory
### Bug report

When trying to mount a configmap with:
```
$ kn service update heloapp -n app --mount /my/path=cm:test
```
it gives error 
```
Error: giving up after 3 retries: admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: invalid value: my.path-1f5d5125: spec.template.spec.volumes[1].name
```

### Expected behavior

It should work and mount the configmap onto `/my/path`

When mounting to folder on root, like `/mypath`, it work:
```
$ kn service update heloapp -n app --mount /mypath=cm:test
Updating Service 'heloapp' in namespace 'app':
...
```

### Steps to reproduce the problem

1. create a configmap called `test`
2. deploy a kn service: `kn service create hello --image gcr.io/cloudrun/hello`
3. use kn service update to mount the configmap to `mypath`: `kn service update hello --mount /my/path=cm:test`

### kn version

```
Version:      v20200728-4c246812
Build Date:   2020-07-28 10:36:25
Git Revision: 4c246812
Supported APIs:
* Serving
  - serving.knative.dev/v1 (knative-serving v0.16.1-0.20200715073232-81d40bfc82a6)
* Eventing
  - sources.knative.dev/v1alpha2 (knative-eventing v0.16.1-0.20200715062032-28f9f61e6131)
  - eventing.knative.dev/v1beta1 (knative-eventing v0.16.1-0.20200715062032-28f9f61e6131)
```

### Knative (serving/eventing) version

> 0.13.2

### PR DESCRIPTION
**Pull request by navidshaikh**
## Description
- Fix client side volume name generation
- Generate volume name compliant with DNS Label names   
  Volume names to follow the DNS label standard as defined in RFC 1123. This means the name must:
  - contain at most 63 characters
  - contain only lowercase alphanumeric characters or '-'
  - start with an alphanumeric character
  - end with an alphanumeric character

## Changes
- Replace non alphanumberic characters with hyphen as the
 volume name must be a DNS_LABEL (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names)
 ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#volume-v1-core
- If replacement generate volume name starting with `-`, prefix `k-`to volume name
- If replacements + checksum suffix, produce volume name longer than 63 chars, trim to 61 characters and append `-n` suffix to it.
## Reference

Fixes #961

### CODE REVIEW

File: pkg/serving/config_changes.go [hunk 1]
```diff
@@ -507,13 +507,13 @@ func GenerateVolumeName(path string) string {
 	builder := &strings.Builder{}
 	for idx, r := range path {
```
Comment (maximilien) [inline pkg/serving/config_changes.go:508]: Where is the code to guard against volume names matching the DNS label name restrictions 1, 3, and 4. That is the bullet points, from spec:

```text
Some resource types require their names to follow the DNS label standard as defined in RFC 1123. This means the name must:

1. contain at most 63 characters
2. contain only lowercase alphanumeric characters or '-'
3. start with an alphanumeric character
4. end with an alphanumeric character
```

There is clearly code to take care of bullet 2 but not the others.

Comment (navidshaikh) [reply pkg/serving/config_changes.go:513]: updated util to generate volume name to adhere to the requirements

**Comment by navidshaikh** [comment]
/retest

       result_collector.go:75: ERROR: Error: ReconcileIngressFailed: Ingress reconciliation failed

**Pull request merged**